File: eu4.c
#include <stdio.h>

// Patch eu4 file so checksum always validates
// Just changing SETE to SETNO
int main(int argc, char **argv) {
  FILE *fp = argc > 1 ? fopen(argv[1], "r+b") : stdin;
  unsigned char buf[16];
  int found1 = 0;
  int found2 = 0;
  int patched = 0;
  long int pos = 0;
  size_t bytes;
  size_t find1[6] = { 0xef, 0xff, 0x85, 0xf6, 0x0f, 0x94 };
  size_t find2[6] = { 0x59, 0x01, 0x85, 0xc0, 0x0f, 0x94 };
  size_t hex;
  size_t i;
  size_t patch[1] = { 0x91 };
  size_t read = sizeof buf;

  if (fp == NULL) {
    fprintf(stderr, "Can't open file at: %s\n", argv[1]);
    return 1;
  }

  while ((bytes = fread(buf, sizeof *buf, read, fp)) == read) {
    for (i = 0; i < read; i++) {
      pos++;

      hex = buf[i];

      if (hex == find1[found1]) {
        found1++;
      } else {
        found1 = 0;
      }

      if (hex == find2[found2]) {
        found2++;
      } else {
        found2 = 0;
      }

      if (found1 == 6 || found2 == 6) {
        patched++;
        printf("Patching 0x%lX\n", pos - i + 2);
        fseek(fp, pos - i + 2, SEEK_SET);
        fwrite(patch, 1, 1, fp);

        if (found1 == 6) {
          found1 = 0;
        } else if (found2 == 6) {
          found2 = 0;
        }
      }
    }
  }

  fclose(fp);

  if (patched != 2) {
    fprintf(stderr, "There should have been 2 offsets found!\n");
    return 1;
  }

  return 0;
}

File: eu4.py
#!/usr/bin/env python3

import re

###              ###
#   EU4 Patcher    #
#  Version 1.28.3  #
###              ###

filename = './eu4'

# Get address of bits
def search(find, s, offset = 0):
    return [(m.start() + offset) for m in re.finditer(find, s)]

def patch(offset, replace):
    with open(filename, 'r+b') as f:
        f.seek(offset)
        f.write(replace)

# Patch checksum, so custom mods do not disable Ironman
def patch_checksum(s):
    # Looking for SETE for checksum validation
    find = b'\x85\xc0\x0f\x94\x45\xef'
    # Replace 4th bit from offset
    offsets = search(find, s, 3)
    if len(offsets) > 0:
        # Replace 94 with 91 (SETE to SETNO)
        for offset in offsets:
            print('Patching checksum offset: {}'.format(hex(offset)))
            patch(offset, b'\x91')
    else:
        # Already patched
        find = b'\x85\xc0\x0f\x91\x45\xef'
        offsets = search(find, s)
        if len(offsets) > 0:
            print('Cannot patch checksum, already patched')
        else:
            print('Cannot patch checksum, unsupported version')

# Patch enable_all_commands
# Still requires any argument to enable_all_commands
def patch_enable_all_commands(s):
    find = b'\x39\xfa\x0f\x85\xd5\x00\x00\x00'
    offsets = search(find, s, 2)
    if len(offsets) > 0:
        # Replace JNZ to JMP
        for offset in offsets:
            print('Patching enable_all_commands offset: {}'.format(hex(offset)))
            patch(offset, b'\xe9\x78\x00\x00\x00\x90')
    else:
        # Already patched
        find = b'\x39\xfa\xe9\x78\x00\x00\x00\x90'
        offsets = search(find, s)
        if len(offsets) > 0:
            print('Cannot patch enable_all_commands, already patched')
        else:
            print('Cannot patch enable_all_commands, unsupported version')

# Decrypt Ironman save file (without 'deiron' command)
def decrypt():
    pass

def main():
    s = None
    with open(filename, 'rb') as f:
        s = f.read()
    if s is None:
        print('Cannot read {}'.format(filename))
        exit(1)
    else:
        patch_checksum(s)
        patch_enable_all_commands(s)

if __name__ == '__main__':
    main()
